File: data structures and algorithms/python/08_025.py
def schedule_tasks(tasks):
   # Sort tasks by priority (descending) and then by deadline
   sorted_tasks = sorted(tasks, 
                       key=lambda x: (-x['priority'], x['deadline']))
   
   # Result list to store tasks in execution order
   result = []
   
   # Timeline to track when resources are allocated
   timeline = {}  # time -> resources being used
   current_time = 0
   
   for task in sorted_tasks:
      task_id = task['id']
      resources = task['resource_needs']
      deadline = task['deadline']
      
      # Find earliest time slot where task can be executed
      execution_time = current_time
      
      # Simple implementation: schedule tasks one after another
      result.append(task_id)
      current_time += resources
   
   return result

# Test with example
tasks = [
   {"id": "T1", "priority": 3, "resource_needs": 5, 
    "deadline": 10},
   {"id": "T2", "priority": 5, "resource_needs": 3, 
    "deadline": 5},
   {"id": "T3", "priority": 2, "resource_needs": 2, 
    "deadline": 7},
   {"id": "T4", "priority": 5, "resource_needs": 1, 
    "deadline": 3}
]

print(schedule_tasks(tasks))  # Output: ["T4", "T2", "T1", "T3"]

File: cpp/08_025.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

struct Task {
    std::string id;
    int priority;
    int resourceNeeds;
    int deadline;
    
    Task(std::string id, int priority, 
         int resourceNeeds, int deadline)
        : id(id), priority(priority), 
          resourceNeeds(resourceNeeds), 
          deadline(deadline) {}
};

std::vector<std::string> scheduleTasks(
        std::vector<Task>& tasks) {
    // Sort by priority (desc) then by deadline (asc)
    std::sort(tasks.begin(), tasks.end(), 
        [](const Task& a, const Task& b) {
            if (a.priority != b.priority) {
                return a.priority > b.priority; // Desc priority
            }
            return a.deadline < b.deadline; // Asc deadline
        });
    
    // Result vector to store tasks in execution order
    std::vector<std::string> result;
    
    // Current time tracker
    int currentTime = 0;
    
    // Schedule each task
    for (const Task& task : tasks) {
        result.push_back(task.id);
        currentTime += task.resourceNeeds;
    }
    
    return result;
}

int main() {
    std::vector<Task> tasks = {
        Task("T1", 3, 5, 10),
        Task("T2", 5, 3, 5),
        Task("T3", 2, 2, 7),
        Task("T4", 5, 1, 3)
    };
    
    std::vector<std::string> schedule = 
        scheduleTasks(tasks);
    
    // Print result
    std::cout << "[";
    for (size_t i = 0; i < schedule.size(); i++) {
        std::cout << schedule[i];
        if (i < schedule.size() - 1) {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl; // Output: [T4, T2, T1, T3]
    
    return 0;
}